{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Team 112\n",
        "## Members:\n",
        "### Tomas Anthony"
      ],
      "metadata": {
        "id": "FVx6yBZyQiJU"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Github Link\n",
        "\n",
        "https://github.com/tomasanthony/cs598-project\n"
      ],
      "metadata": {
        "id": "-CfIGmp3RecR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Video Presentation\n",
        "https://youtu.be/ERturLrhywY"
      ],
      "metadata": {
        "id": "KZFafTVzXlLu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introduction\n",
        "\n",
        "## Problem\n",
        "\n",
        "Alzheimer’s disease is a a prominent brain disease that is the fifth-leading cause of death among\n",
        "individuals aged 65 and older, according to data from 2021 (Association, 2023); It is estimated\n",
        "that 10% of the 65 and older age group are suffering from Alzheimer’s and 8-11% may have Mild\n",
        "Cognitive Impairment (MCI), a precursor to Alzheimer’s (Association, 2023). Alzheimer’s and MCI\n",
        "are also currently under-diagnosed in the primary care setting, with people in the earlier stages of the\n",
        "disease (MCI) especially going under-diagnosed (Association, 2023). In the paper Golovanevsky\n",
        "et al. (2022), researchers noted the difficulty in diagnosing Alzheimer’s disease and emphasized in their research the potential to aid medical professionals with machine learning.\n",
        "\n",
        "## Original Paper\n",
        "\n",
        "The original paper that is replicated here is titled ”Multimodal Attention-based\n",
        "Deep Learning for Alzheimer’s Disease Diagnosis” and was published in the Journal of the American Medical Informatics Association in 2022 (Golovanevsky et al., 2022).\n",
        "\n",
        "### Paper Methodologies\n",
        "\n",
        "#### Attention Layers\n",
        "The approach used by Golovanevsky et al. involved a multi-modal deep learning model with multiple attention layers. The multi-modal approach mentioned in the title involved ingesting different\n",
        "modalities of data - imaging, clinical, and genetic data - and using the attention layers to preserve inter-modal interactions in the data.\n",
        "\n",
        "The researchers followed the example set in the original Transformers paper and used self-attention and\n",
        "cross-attention mechanisms for their attention architecture.\n",
        "\n",
        "#### Fully Connected Networks\n",
        "Each modality of data was initially fed through a fully connected neural network with each of these neural networks having its own hyper-parameter tuning\n",
        "done. The output of the fully-connected neural networks was then fed into attention-layers.\n",
        "\n",
        "#### Output Layer\n",
        "\n",
        "The output of the cross-attention layer\n",
        "is fed into a final fully connected layer. The performance of the model produced at the end of the\n",
        "pipeline was used to select the tune parameters and rerun the pipeline, and the best parameters for\n",
        "the model were used (Golovanevsky et al., 2022).\n",
        "\n",
        "#### Data\n",
        "The data used by Golovanevsky et al. was comprised of the ADNI1, ADNI2, and GO\n",
        "1\n",
        "datasets from the Alzheimer’s Disease Neuroimaging Initiative (ADNI) database.\n",
        "\n",
        "\n",
        "## Current State of the Art\n",
        "\n",
        "The MADDi model introduced by Golovanevsky et al. (2022) achieved the current state of the art results. The methods used include :\n",
        "- **Multiple Attention Layers:** Utilizing both self-attention and cross-attention layers, an approach that was modeled after the original Transformers paper.\n",
        "- **Multi-modal Data:** The use of multi-modal data was a defining characteristic of the approach. Clinical, genetic, and imaging data were all used to train the model.\n",
        "- **Hyper-parameter Tuning:** Each data modality was initially fed through a fully connected neural network, whose parameters were defined using extensive hyper-parameter tuning.\n",
        "\n",
        "The MADDi model improved upon the previous state of the art F1-Score and Accuracy by 4% and 10.8%, respectively.\n",
        "\n",
        "## Innovations\n",
        "\n",
        "- **Cross-Modal Attention:** The primary innovation was the extensive use of attention layers. Each data modality is included in a cross-attention layer, in addition to their own self-attention layers. This approach captures the significance of cross-modality interactions in the data. The researches noted that previously, multi-modal approaches simply concatenated the different data instead of utilizing the multiple modalities as distinct data types.\n",
        "\n",
        "## Contributions\n",
        "- integrating multimodal inputs, multi-task\n",
        "classification, and cross-modal attention for capturing interactions\n",
        "- **Multimodal Inputs:** The paper introduced the use of multimodal inputs, a distinction from using multimodal data that is concatenated into a single input.\n",
        "- **Multi-task Classification:** Other approaches did not use multi-task classification to address control, mild cognitiive impairment, and Alzheimer's diagnosis at in a single model.\n",
        "- **Cross-Modal Attention:** As noted previously, the use of cross-modal attention and transformers architecture was a key contribution to the field.\n",
        "\n",
        "## Performance\n",
        "- **F1-Score:** 91.41%\n",
        "- **Accuracy:** 96.88%\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "MQ0sNuMePBXx"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Scope of Reproducibility:\n",
        "\n",
        "1. State of the art results for Accuracy, Precision, Recall, and\n",
        "F1-Scores averaged across three classes (Control, Moderate Cognitive Impairment, and Alzheimer’s\n",
        "disease).\n",
        "2. Elevated significance of clinical data compared to other data modalities.\n",
        "3. Multi-modal models utilizing all three modalities outperform uni-modal and dual-modality models\n",
        "4. Inclusion of additional, more recent data in\n",
        "the ADNI database, specifically ADNI3, will lead to performance improvements.\n"
      ],
      "metadata": {
        "id": "uygL9tTPSVHB"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Methodology\n",
        "The existing code for the MADDi model and the orginal paper was used and adjusted for this project\n",
        "\n"
      ],
      "metadata": {
        "id": "xWAHJ_1CdtaA"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Environment\n"
      ],
      "metadata": {
        "id": "-UB3zvgEZCj3"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Python Version\n",
        "This project was developed using **Python 3.10**.\n",
        "\n",
        "The MADDi model can be replicated with Python version greater than **3.7**.\n"
      ],
      "metadata": {
        "id": "EmDROSrUZMim"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Dependencies\n",
        "\n",
        "#### Notebook\n",
        "The notebook primarily uses **tensorflow**, **numpy**, **pandas**, **sklearn**, and **matplotlib**.\n",
        "\n",
        "To fully replicate the original paper, a ``` requirements.txt ``` file is provided [here](https://github.com/tomasanthony/cs598-project/blob/main/general/requirements.txt).\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "9SUz-leaZfeF"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##  Data\n",
        "Data includes ADNI 1, 2, 3, and GO datasets.\n",
        " ### Data Source\n",
        "  The use of the ADNI datasets require requesting access. The data is stored in the Image and Data\n",
        "Archive (IDA). The access request involves providing information about the intended use of the\n",
        "project and the researchers involved. Access to the datasets has already been granted to this project’s\n",
        "participants.\n",
        "\n",
        "### Data Download Instructions\n",
        "\n",
        "To access the ADNI dataset, follow these steps:\n",
        "\n",
        "\n",
        "1.   **Complete the Data Use Agreement form**: The  [ADNI Data Use Agreement](https://ida.loni.usc.edu/collaboration/access/appLicense.jsp) can be accessed directly.\n",
        "2.   **Apply for Access**: Submit an application for accessing the ADNI dataset. This includes providing your affiliation and the use case for the data.\n",
        "3. **Download Data**: Once the access request is approved. Users can log in to the LONI website to access and download ADNI datasets. Include study, imaging, and genetic data.\n",
        "\n",
        "### Data Description\n",
        "\n",
        "The dataset used in this study comprises multi-modal data sourced from ADNI (Alzheimer's Disease Neuroimaging Initiative). Given the confidential nature of this data, detailed charts and visualizations cannot be provided. Below is a general description of the types of data included, based on the methodologies outlined in the MADDi paper:\n",
        "\n",
        "1. **Imaging Data**:\n",
        "   - Consists of cross-sectional MRI scans from the baseline screenings of participants.\n",
        "   - Images have been standardized by ADNI with GradWarp, B1 Correction, and N3 preprocessing steps.\n",
        "   - Three slices of MRI images are included per patient, resized to 72x72 pixels.\n",
        "\n",
        "2. **Genetic Data**:\n",
        "   - Consists of whole genome sequencing (WGS) performed by Illumina's non-Clinical Laboratory Improvement Amendments (non-CLIA) process.\n",
        "   - The initial dataset is reduced down to approximately 15,000 SNPs during preprocessing.\n",
        "   - Data is in the form of large VCF files curated by ADNI in 2014.\n",
        "\n",
        "3. **Clinical Data**:\n",
        "   - Consists of initial assessments from neurological exams, cognitive tests, and demographic information.\n",
        "   - 29 features\n",
        "\n",
        "### Dataset Overlap\n",
        "\n",
        "The final dataset included an overlap of 239 patients.\n",
        "\n",
        "### Note on non-public nature of the datasets\n",
        "\n",
        "The datasets used in this project are non-public and the reproduction and visualization of data samples will not be done to preserve their confidential nature.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "n9Qv0eRJzr6v"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "data = {\n",
        "    \"Total Participants\": [2384, 551, 805, 239],\n",
        "    \"Control\": [942, 278, 241, 165],\n",
        "    \"Mild Cognitive Impairment\": [796, 123, 318, 39],\n",
        "    \"Alzheimer's Disease\": [646, 150, 246, 35]\n",
        "}\n",
        "\n",
        "index_labels = ['Clinical', 'Imaging', 'Genetic', 'Overlap']\n",
        "\n",
        "df = pd.DataFrame(data, index=index_labels)\n",
        "\n",
        "df\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 175
        },
        "id": "X_fkSbcf3o3o",
        "outputId": "9d4ce4ab-a370-44df-cd69-ad7010f7d988"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          Total Participants  Control  Mild Cognitive Impairment  \\\n",
              "Clinical                2384      942                        796   \n",
              "Imaging                  551      278                        123   \n",
              "Genetic                  805      241                        318   \n",
              "Overlap                  239      165                         39   \n",
              "\n",
              "          Alzheimer's Disease  \n",
              "Clinical                  646  \n",
              "Imaging                   150  \n",
              "Genetic                   246  \n",
              "Overlap                    35  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-4ff3a236-07b4-4203-960e-4c5d0957f70c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Total Participants</th>\n",
              "      <th>Control</th>\n",
              "      <th>Mild Cognitive Impairment</th>\n",
              "      <th>Alzheimer's Disease</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Clinical</th>\n",
              "      <td>2384</td>\n",
              "      <td>942</td>\n",
              "      <td>796</td>\n",
              "      <td>646</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Imaging</th>\n",
              "      <td>551</td>\n",
              "      <td>278</td>\n",
              "      <td>123</td>\n",
              "      <td>150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Genetic</th>\n",
              "      <td>805</td>\n",
              "      <td>241</td>\n",
              "      <td>318</td>\n",
              "      <td>246</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Overlap</th>\n",
              "      <td>239</td>\n",
              "      <td>165</td>\n",
              "      <td>39</td>\n",
              "      <td>35</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4ff3a236-07b4-4203-960e-4c5d0957f70c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-4ff3a236-07b4-4203-960e-4c5d0957f70c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-4ff3a236-07b4-4203-960e-4c5d0957f70c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-5174ac3a-3b57-43f3-87c8-903043fb01b8\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5174ac3a-3b57-43f3-87c8-903043fb01b8')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-5174ac3a-3b57-43f3-87c8-903043fb01b8 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 4,\n  \"fields\": [\n    {\n      \"column\": \"Total Participants\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 954,\n        \"min\": 239,\n        \"max\": 2384,\n        \"num_unique_values\": 4,\n        \"samples\": [\n          551,\n          239,\n          2384\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Control\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 360,\n        \"min\": 165,\n        \"max\": 942,\n        \"num_unique_values\": 4,\n        \"samples\": [\n          278,\n          165,\n          942\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Mild Cognitive Impairment\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 338,\n        \"min\": 39,\n        \"max\": 796,\n        \"num_unique_values\": 4,\n        \"samples\": [\n          123,\n          39,\n          796\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Alzheimer's Disease\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 265,\n        \"min\": 35,\n        \"max\": 646,\n        \"num_unique_values\": 4,\n        \"samples\": [\n          150,\n          35,\n          646\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Preprocessing the data\n",
        "The preprocessing code for each modalities can be found in Github.\n",
        "\n",
        "1. [Initial Preparation](https://github.com/tomasanthony/cs598-project/blob/main/general/diagnosis_making.ipynb)\n",
        "2. [Clinical Preprocessing](https://github.com/tomasanthony/cs598-project/tree/main/preprocess_clinical)\n",
        "3. [Genetic Preprocessing](https://github.com/tomasanthony/cs598-project/tree/main/preprocess_genetic)\n",
        "4. [Image Preprocessing](https://github.com/tomasanthony/cs598-project/tree/main/preprocess_images)\n",
        "\n",
        "The final preprocessing step, which involves merging the datasets, is included below.\n"
      ],
      "metadata": {
        "id": "2FFD0u4N4wZK"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BZScZNbROw-N"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "# dir and function to load raw data\n",
        "raw_data_dir = '/content/gdrive/My Drive/Colab Notebooks/DL4H Project/'\n",
        "\n",
        "clinical_data_dir = raw_data_dir + \"clinical_data/\"\n",
        "image_data_dir = raw_data_dir + \"image_data/\"\n",
        "image_data_dir = raw_data_dir + \"genetic_data/\"\n",
        "\n",
        "\n",
        "def load_raw_data(raw_data_dir):\n",
        "  vcf = pd.read_pickle(\"all_vcfs.pkl\")\n",
        "  c = pd.read_csv(\"clinical.csv\").drop(\"Unnamed: 0\", axis=1).rename(columns={\"PTID\":\"subject\"})\n",
        "  img = pd.read_pickle('mri_meta.pkl')[[\"img_array\", \"subject\", \"label\"]]\n",
        "  return vcf, c, img\n",
        "\n",
        "vcf, c, img = load_raw_data(raw_data_dir)\n",
        "\n",
        "# Merge modalities and calculate/merge stats\n",
        "def merge_modalities(vcf, c, img):\n",
        "  vcf.head()\n",
        "  c.head()\n",
        "  c = c.rename(columns = {\"Group\":\"GROUP\"})\n",
        "  a = vcf.merge(c, on = [\"subject\", \"GROUP\"]).merge(img, on = \"subject\")\n",
        "\n",
        "merged_data = merge_modalities(vcf, c, img)\n",
        "\n",
        "def calculate_stats(a)\n",
        "  subject_counts = a[\"subject\"].value_counts()\n",
        "  group_counts = a[\"GROUP\"].value_counts()\n",
        "\n",
        "  return subject_counts, group_counts\n",
        "\n",
        "# process raw data\n",
        "def process_data(a):\n",
        "  # Set columns for aggregated data\n",
        "  cols = list(set(a.columns) - set([\"PTID\", \"label\", \"GROUP\",\n",
        "                                  \"RID\", \"ID\", \"Group\", \"Phase\", \"SITEID\", \"VISCODE\", \"VISCODE2\", \"USERDATE\", \"USERDATE2\", \"update_stamp\", \"DX\", \"Unnamed: 0\"]))\n",
        "\n",
        "  X= a[cols]\n",
        "  y = a[\"GROUP\"]\n",
        "  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, stratify=y)\n",
        "  X_test[[\"subject\"]].to_csv(\"overlap_test_set.csv\")\n",
        "  snp_cols = set(X_train.columns).intersection(set(vcf.columns))\n",
        "  X_train_snp = X_train[snp_cols]\n",
        "  X_test_snp = X_test[snp_cols]\n",
        "  img_cols = set(X_train.columns).intersection(set(img.columns))\n",
        "  print(len(img.columns))\n",
        "  print(len(img_cols))\n",
        "  X_train_img = X_train[img_cols]\n",
        "  X_test_img = X_test[img_cols]\n",
        "  clin_cols = set(X_train.columns).intersection(set(c.columns))\n",
        "  print(len(c.columns))\n",
        "  print(len(clin_cols))\n",
        "  X_train_clin = X_train[clin_cols]\n",
        "  X_test_clin = X_test[clin_cols]\n",
        "\n",
        "  # Define the directory in Google Drive where the output files will be stored\n",
        "  output_pickle_path = raw_data_dir + \"processed_pickle_data/\"\n",
        "\n",
        "  # Converting arrays to pandas DataFrame and saving them as Pickle files\n",
        "  # This is useful for data persistence, making it easier to load preprocessed data without re-running the preprocessing steps\n",
        "\n",
        "  # SNP datasets\n",
        "  pd.DataFrame(X_train_snp).to_pickle(output_pickle_path + \"X_train_snp.pkl\")\n",
        "  pd.DataFrame(X_test_snp).to_pickle(output_pickle_path + \"X_test_snp.pkl\")\n",
        "\n",
        "  # Labels\n",
        "  pd.DataFrame(y_train).to_pickle(output_pickle_path + \"y_train.pkl\")\n",
        "  pd.DataFrame(y_test).to_pickle(output_pickle_path + \"y_test.pkl\")\n",
        "\n",
        "  # Clinical data\n",
        "  pd.DataFrame(X_train_clin).to_pickle(output_pickle_path + \"X_train_clinical.pkl\")\n",
        "  pd.DataFrame(X_test_clin).to_pickle(output_pickle_path + \"X_test_clinical.pkl\")\n",
        "\n",
        "  # Imaging data\n",
        "  pd.DataFrame(X_train_img).to_pickle(output_pickle_path + \"X_train_img.pkl\")\n",
        "  pd.DataFrame(X_test_img).to_pickle(output_pickle_path + \"X_test_img.pkl\")\n",
        "\n",
        "processed_data = process_data(merged_data)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##   Model\n",
        "\n",
        "### Original Paper\n",
        "[”Multimodal Attention-based Deep Learning for Alzheimer’s Disease Diagnosis\"](https://arxiv.org/abs/2206.08826),  Michal Golovanevsky, Carsten Eickhoff, Ritambhara Singh.\n",
        "\n",
        "[Original Paper Github](https://github.com/rsinghlab/MADDi/tree/main)\n",
        "\n",
        "### Description\n",
        "The model uses a multi-modal, attention-based approach. It incorporates three distinct data modalities from the ADNI dataset - imaging, clinical, and genetic.\n",
        "\n",
        "These three modalities serve as inputs to the self-attention, and then cross-attention layers where the modalities are integrated together. They are then fed into a fully-connected layer for final output.\n",
        "\n",
        "- **Imaging Input**: This modality uses CNN layers to process the baseline screening MRI images for study participants. The CNN layers are then flattened and fed through a dense layer.\n",
        "- **Clinical Input**: This modality uses dense layers, which are then normalized and fed through dropout layers.\n",
        "- **Genetic Pathway**: This modality also uses dense layers, followed by normalization and dropout layers\n",
        "\n",
        "### Attention Architecture\n",
        "- **Self-Attention**: Each modality is fed to a self-attention layer\n",
        "- **Cross-Modal Attention**: The cross-attention layer integrates the separate modalities and captures interactions between themm\n"
      ],
      "metadata": {
        "id": "3muyDPFPbozY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import random\n",
        "import gc\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "from sklearn.utils import compute_class_weight\n",
        "from sklearn.metrics import classification_report, precision_recall_curve, precision_recall_fscore_support\n",
        "import tensorflow as tf\n",
        "from tensorflow.keras.models import Model, Sequential\n",
        "from tensorflow.keras.layers import Input, Dense, Dropout, Flatten, BatchNormalization, Conv2D, MultiHeadAttention, concatenate\n",
        "from tensorflow.keras.optimizers import Adam\n",
        "from tensorflow.keras.utils import to_categorical\n",
        "\n",
        "# Configure TensorFlow session\n",
        "config = tf.compat.v1.ConfigProto()\n",
        "config.gpu_options.allow_growth = True\n",
        "sess = tf.compat.v1.Session(config=config)\n"
      ],
      "metadata": {
        "id": "MT4Ag-CEIEsh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def make_img(t_img):\n",
        "    \"\"\" Load image data from pickle and prepare for training. \"\"\"\n",
        "    img = pd.read_pickle(t_img)\n",
        "    img_l = [img.values[i][0] for i in range(len(img))]\n",
        "    return np.array(img_l)\n",
        "\n",
        "def reset_random_seeds(seed):\n",
        "    \"\"\" Fix random seed for reproducibility. \"\"\"\n",
        "    os.environ['PYTHONHASHSEED'] = str(seed)\n",
        "    tf.random.set_seed(seed)\n",
        "    np.random.seed(seed)\n",
        "    random.seed(seed)"
      ],
      "metadata": {
        "id": "zeD70U-tIyZb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def create_model_snp():\n",
        "    \"\"\" Build and return a Sequential model for SNP data. \"\"\"\n",
        "    model = Sequential([\n",
        "        Input(shape=(15965,)),  # Input layer with shape = the number of SNPs used as features rom the genetic dataset\n",
        "        Dense(200, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.5),\n",
        "        Dense(100, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.3),\n",
        "        Dense(50, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.2)\n",
        "    ])\n",
        "    return model\n",
        "\n",
        "def create_model_clinical():\n",
        "    \"\"\" Build and return a Sequential model for clinical data. \"\"\"\n",
        "    model = Sequential([\n",
        "        Input(shape=(185,)),  # Input layer with shape = the number of features in the clinical dataset\n",
        "        Dense(200, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.5),\n",
        "        Dense(100, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.3),\n",
        "        Dense(50, activation=\"relu\"),\n",
        "        BatchNormalization(),\n",
        "        Dropout(0.2)\n",
        "    ])\n",
        "    return model\n",
        "\n",
        "def create_model_img():\n",
        "    \"\"\" Build and return a Sequential model for imaging data. \"\"\"\n",
        "    model = Sequential([\n",
        "        Input(shape=(72, 72, 3)),  # Shape = height, width, number of slices from the MRI screenings images of ADNI1 dataset\n",
        "        Conv2D(72, (3, 3), activation='relu'),\n",
        "        Conv2D(64, (3, 3), activation='relu'),\n",
        "        Conv2D(32, (3, 3), activation='relu'),\n",
        "        Flatten(),\n",
        "        Dense(50, activation='relu')\n",
        "    ])\n",
        "    return model"
      ],
      "metadata": {
        "id": "Nmiu1xxbIzLp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from tensorflow.keras.utils import model_to_dot\n",
        "from IPython.display import HTML\n",
        "\n",
        "def get_svg(model):\n",
        "    return model_to_dot(model, show_shapes=True, show_layer_names=True, dpi=90).create(prog='dot', format='svg').decode('utf-8')\n",
        "\n",
        "\n",
        "\n",
        "model_snp = create_model_snp()\n",
        "model_clinical = create_model_clinical()\n",
        "model_img = create_model_img()\n",
        "\n",
        "print(\"SNP Model Summary:\")\n",
        "model_snp.summary()\n",
        "print(\"\\nClinical Model Summary:\")\n",
        "model_clinical.summary()\n",
        "print(\"\\nImaging Model Summary:\")\n",
        "model_img.summary()\n",
        "\n",
        "model_snp_svg = get_svg(model_snp)\n",
        "model_clinical_svg = get_svg(model_clinical)\n",
        "model_img_svg = get_svg(model_img)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LV1ELXD4ZwQF",
        "outputId": "491124a6-0beb-49d3-ec2b-ea452f7250dc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "SNP Model Summary:\n",
            "Model: \"sequential_32\"\n",
            "_________________________________________________________________\n",
            " Layer (type)                Output Shape              Param #   \n",
            "=================================================================\n",
            " dense_76 (Dense)            (None, 200)               3193200   \n",
            "                                                                 \n",
            " batch_normalization_66 (Ba  (None, 200)               800       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_66 (Dropout)        (None, 200)               0         \n",
            "                                                                 \n",
            " dense_77 (Dense)            (None, 100)               20100     \n",
            "                                                                 \n",
            " batch_normalization_67 (Ba  (None, 100)               400       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_67 (Dropout)        (None, 100)               0         \n",
            "                                                                 \n",
            " dense_78 (Dense)            (None, 50)                5050      \n",
            "                                                                 \n",
            " batch_normalization_68 (Ba  (None, 50)                200       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_68 (Dropout)        (None, 50)                0         \n",
            "                                                                 \n",
            "=================================================================\n",
            "Total params: 3219750 (12.28 MB)\n",
            "Trainable params: 3219050 (12.28 MB)\n",
            "Non-trainable params: 700 (2.73 KB)\n",
            "_________________________________________________________________\n",
            "\n",
            "Clinical Model Summary:\n",
            "Model: \"sequential_33\"\n",
            "_________________________________________________________________\n",
            " Layer (type)                Output Shape              Param #   \n",
            "=================================================================\n",
            " dense_79 (Dense)            (None, 200)               37200     \n",
            "                                                                 \n",
            " batch_normalization_69 (Ba  (None, 200)               800       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_69 (Dropout)        (None, 200)               0         \n",
            "                                                                 \n",
            " dense_80 (Dense)            (None, 100)               20100     \n",
            "                                                                 \n",
            " batch_normalization_70 (Ba  (None, 100)               400       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_70 (Dropout)        (None, 100)               0         \n",
            "                                                                 \n",
            " dense_81 (Dense)            (None, 50)                5050      \n",
            "                                                                 \n",
            " batch_normalization_71 (Ba  (None, 50)                200       \n",
            " tchNormalization)                                               \n",
            "                                                                 \n",
            " dropout_71 (Dropout)        (None, 50)                0         \n",
            "                                                                 \n",
            "=================================================================\n",
            "Total params: 63750 (249.02 KB)\n",
            "Trainable params: 63050 (246.29 KB)\n",
            "Non-trainable params: 700 (2.73 KB)\n",
            "_________________________________________________________________\n",
            "\n",
            "Imaging Model Summary:\n",
            "Model: \"sequential_34\"\n",
            "_________________________________________________________________\n",
            " Layer (type)                Output Shape              Param #   \n",
            "=================================================================\n",
            " conv2d_30 (Conv2D)          (None, 70, 70, 72)        2016      \n",
            "                                                                 \n",
            " conv2d_31 (Conv2D)          (None, 68, 68, 64)        41536     \n",
            "                                                                 \n",
            " conv2d_32 (Conv2D)          (None, 66, 66, 32)        18464     \n",
            "                                                                 \n",
            " flatten_10 (Flatten)        (None, 139392)            0         \n",
            "                                                                 \n",
            " dense_82 (Dense)            (None, 50)                6969650   \n",
            "                                                                 \n",
            "=================================================================\n",
            "Total params: 7031666 (26.82 MB)\n",
            "Trainable params: 7031666 (26.82 MB)\n",
            "Non-trainable params: 0 (0.00 Byte)\n",
            "_________________________________________________________________\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "html = f\"\"\"\n",
        "<div style='width: 100%; display: flex;'>\n",
        "    <div style='width: 33%; text-align: center;'>\n",
        "        <h2>SNP Model</h2>\n",
        "        <div>{model_snp_svg}</div>\n",
        "    </div>\n",
        "    <div style='width: 33%; text-align: center;'>\n",
        "        <h2>Clinical Model</h2>\n",
        "        <div>{model_clinical_svg}</div>\n",
        "    </div>\n",
        "    <div style='width: 33%; text-align: center;'>\n",
        "        <h2>Imaging Model</h2>\n",
        "        <div>{model_img_svg}</div>\n",
        "    </div>\n",
        "</div>\n",
        "\"\"\"\n",
        "HTML(html)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "7T4EAx5_chxo",
        "outputId": "086b6161-235b-4909-efae-acb7936828e2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "<div style='width: 100%; display: flex;'>\n",
              "    <div style='width: 33%; text-align: center;'>\n",
              "        <h2>SNP Model</h2>\n",
              "        <div><?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
              "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
              " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
              "<!-- Generated by graphviz version 2.43.0 (0)\n",
              " -->\n",
              "<!-- Title: G Pages: 1 -->\n",
              "<svg width=\"369pt\" height=\"1003pt\"\n",
              " viewBox=\"0.00 0.00 295.00 802.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
              "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.8 0.8) rotate(0) translate(4 798)\">\n",
              "<title>G</title>\n",
              "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-798 291,-798 291,4 -4,4\"/>\n",
              "<!-- 139796614532256 -->\n",
              "<g id=\"node1\" class=\"node\">\n",
              "<title>139796614532256</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"25,-747.5 25,-793.5 262,-793.5 262,-747.5 25,-747.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"63.5\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">input_28</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"25,-770.5 102,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"63.5\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">InputLayer</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"102,-747.5 102,-793.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"129.5\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"102,-770.5 157,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"129.5\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"157,-747.5 157,-793.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"209.5\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 15965)]</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"157,-770.5 262,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"209.5\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 15965)]</text>\n",
              "</g>\n",
              "<!-- 139796614535952 -->\n",
              "<g id=\"node2\" class=\"node\">\n",
              "<title>139796614535952</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"34,-664.5 34,-710.5 253,-710.5 253,-664.5 34,-664.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"68\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_76</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"34,-687.5 102,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"68\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"102,-664.5 102,-710.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"129.5\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"102,-687.5 157,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"129.5\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"157,-664.5 157,-710.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 15965)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"157,-687.5 253,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796614532256&#45;&gt;139796614535952 -->\n",
              "<g id=\"edge1\" class=\"edge\">\n",
              "<title>139796614532256&#45;&gt;139796614535952</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-747.37C143.5,-739.15 143.5,-729.66 143.5,-720.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-720.61 143.5,-710.61 140,-720.61 147,-720.61\"/>\n",
              "</g>\n",
              "<!-- 139796614535184 -->\n",
              "<g id=\"node3\" class=\"node\">\n",
              "<title>139796614535184</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"0,-581.5 0,-627.5 287,-627.5 287,-581.5 0,-581.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_66</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"0,-604.5 149,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-581.5 149,-627.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-604.5 204,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-581.5 204,-627.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-604.5 287,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796614535952&#45;&gt;139796614535184 -->\n",
              "<g id=\"edge2\" class=\"edge\">\n",
              "<title>139796614535952&#45;&gt;139796614535184</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-664.37C143.5,-656.15 143.5,-646.66 143.5,-637.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-637.61 143.5,-627.61 140,-637.61 147,-637.61\"/>\n",
              "</g>\n",
              "<!-- 139796614534560 -->\n",
              "<g id=\"node4\" class=\"node\">\n",
              "<title>139796614534560</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"35,-498.5 35,-544.5 252,-544.5 252,-498.5 35,-498.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_66</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"35,-521.5 114,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-498.5 114,-544.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-521.5 169,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-498.5 169,-544.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-521.5 252,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796614535184&#45;&gt;139796614534560 -->\n",
              "<g id=\"edge3\" class=\"edge\">\n",
              "<title>139796614535184&#45;&gt;139796614534560</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-581.37C143.5,-573.15 143.5,-563.66 143.5,-554.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-554.61 143.5,-544.61 140,-554.61 147,-554.61\"/>\n",
              "</g>\n",
              "<!-- 139796614536624 -->\n",
              "<g id=\"node5\" class=\"node\">\n",
              "<title>139796614536624</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"40.5,-415.5 40.5,-461.5 246.5,-461.5 246.5,-415.5 40.5,-415.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_77</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"40.5,-438.5 108.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-415.5 108.5,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-438.5 163.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-415.5 163.5,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-438.5 246.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796614534560&#45;&gt;139796614536624 -->\n",
              "<g id=\"edge4\" class=\"edge\">\n",
              "<title>139796614534560&#45;&gt;139796614536624</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-498.37C143.5,-490.15 143.5,-480.66 143.5,-471.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-471.61 143.5,-461.61 140,-471.61 147,-471.61\"/>\n",
              "</g>\n",
              "<!-- 139796614542816 -->\n",
              "<g id=\"node6\" class=\"node\">\n",
              "<title>139796614542816</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"0,-332.5 0,-378.5 287,-378.5 287,-332.5 0,-332.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_67</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"0,-355.5 149,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-332.5 149,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-355.5 204,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-332.5 204,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-355.5 287,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796614536624&#45;&gt;139796614542816 -->\n",
              "<g id=\"edge5\" class=\"edge\">\n",
              "<title>139796614536624&#45;&gt;139796614542816</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-415.37C143.5,-407.15 143.5,-397.66 143.5,-388.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-388.61 143.5,-378.61 140,-388.61 147,-388.61\"/>\n",
              "</g>\n",
              "<!-- 139796614532688 -->\n",
              "<g id=\"node7\" class=\"node\">\n",
              "<title>139796614532688</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"35,-249.5 35,-295.5 252,-295.5 252,-249.5 35,-249.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_67</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"35,-272.5 114,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-249.5 114,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-272.5 169,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-249.5 169,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-272.5 252,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796614542816&#45;&gt;139796614532688 -->\n",
              "<g id=\"edge6\" class=\"edge\">\n",
              "<title>139796614542816&#45;&gt;139796614532688</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-332.37C143.5,-324.15 143.5,-314.66 143.5,-305.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-305.61 143.5,-295.61 140,-305.61 147,-305.61\"/>\n",
              "</g>\n",
              "<!-- 139796614544352 -->\n",
              "<g id=\"node8\" class=\"node\">\n",
              "<title>139796614544352</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"40.5,-166.5 40.5,-212.5 246.5,-212.5 246.5,-166.5 40.5,-166.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_78</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"40.5,-189.5 108.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-166.5 108.5,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-189.5 163.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-166.5 163.5,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-189.5 246.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614532688&#45;&gt;139796614544352 -->\n",
              "<g id=\"edge7\" class=\"edge\">\n",
              "<title>139796614532688&#45;&gt;139796614544352</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-249.37C143.5,-241.15 143.5,-231.66 143.5,-222.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-222.61 143.5,-212.61 140,-222.61 147,-222.61\"/>\n",
              "</g>\n",
              "<!-- 139796614541760 -->\n",
              "<g id=\"node9\" class=\"node\">\n",
              "<title>139796614541760</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"3.5,-83.5 3.5,-129.5 283.5,-129.5 283.5,-83.5 3.5,-83.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_68</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"3.5,-106.5 152.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"152.5,-83.5 152.5,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"180\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"152.5,-106.5 207.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"180\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"207.5,-83.5 207.5,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"207.5,-106.5 283.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614544352&#45;&gt;139796614541760 -->\n",
              "<g id=\"edge8\" class=\"edge\">\n",
              "<title>139796614544352&#45;&gt;139796614541760</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-166.37C143.5,-158.15 143.5,-148.66 143.5,-139.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-139.61 143.5,-129.61 140,-139.61 147,-139.61\"/>\n",
              "</g>\n",
              "<!-- 139796614534464 -->\n",
              "<g id=\"node10\" class=\"node\">\n",
              "<title>139796614534464</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"38.5,-0.5 38.5,-46.5 248.5,-46.5 248.5,-0.5 38.5,-0.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_68</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"38.5,-23.5 117.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"117.5,-0.5 117.5,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"145\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"117.5,-23.5 172.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"145\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"172.5,-0.5 172.5,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"172.5,-23.5 248.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614541760&#45;&gt;139796614534464 -->\n",
              "<g id=\"edge9\" class=\"edge\">\n",
              "<title>139796614541760&#45;&gt;139796614534464</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-83.37C143.5,-75.15 143.5,-65.66 143.5,-56.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-56.61 143.5,-46.61 140,-56.61 147,-56.61\"/>\n",
              "</g>\n",
              "</g>\n",
              "</svg>\n",
              "</div>\n",
              "    </div>\n",
              "    <div style='width: 33%; text-align: center;'>\n",
              "        <h2>Clinical Model</h2>\n",
              "        <div><?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
              "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
              " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
              "<!-- Generated by graphviz version 2.43.0 (0)\n",
              " -->\n",
              "<!-- Title: G Pages: 1 -->\n",
              "<svg width=\"369pt\" height=\"1003pt\"\n",
              " viewBox=\"0.00 0.00 295.00 802.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
              "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.8 0.8) rotate(0) translate(4 798)\">\n",
              "<title>G</title>\n",
              "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-798 291,-798 291,4 -4,4\"/>\n",
              "<!-- 139796614539024 -->\n",
              "<g id=\"node1\" class=\"node\">\n",
              "<title>139796614539024</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"31.5,-747.5 31.5,-793.5 255.5,-793.5 255.5,-747.5 31.5,-747.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"70\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">input_29</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"31.5,-770.5 108.5,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"70\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">InputLayer</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-747.5 108.5,-793.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-770.5 163.5,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-747.5 163.5,-793.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"209.5\" y=\"-778.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 185)]</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-770.5 255.5,-770.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"209.5\" y=\"-755.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 185)]</text>\n",
              "</g>\n",
              "<!-- 139796616574240 -->\n",
              "<g id=\"node2\" class=\"node\">\n",
              "<title>139796616574240</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"40.5,-664.5 40.5,-710.5 246.5,-710.5 246.5,-664.5 40.5,-664.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_79</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"40.5,-687.5 108.5,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-664.5 108.5,-710.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-687.5 163.5,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-664.5 163.5,-710.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-695.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 185)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-687.5 246.5,-687.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-672.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796614539024&#45;&gt;139796616574240 -->\n",
              "<g id=\"edge1\" class=\"edge\">\n",
              "<title>139796614539024&#45;&gt;139796616574240</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-747.37C143.5,-739.15 143.5,-729.66 143.5,-720.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-720.61 143.5,-710.61 140,-720.61 147,-720.61\"/>\n",
              "</g>\n",
              "<!-- 139796616604944 -->\n",
              "<g id=\"node3\" class=\"node\">\n",
              "<title>139796616604944</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"0,-581.5 0,-627.5 287,-627.5 287,-581.5 0,-581.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_69</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"0,-604.5 149,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-581.5 149,-627.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-604.5 204,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-581.5 204,-627.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-612.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-604.5 287,-604.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-589.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796616574240&#45;&gt;139796616604944 -->\n",
              "<g id=\"edge2\" class=\"edge\">\n",
              "<title>139796616574240&#45;&gt;139796616604944</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-664.37C143.5,-656.15 143.5,-646.66 143.5,-637.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-637.61 143.5,-627.61 140,-637.61 147,-637.61\"/>\n",
              "</g>\n",
              "<!-- 139796616605184 -->\n",
              "<g id=\"node4\" class=\"node\">\n",
              "<title>139796616605184</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"35,-498.5 35,-544.5 252,-544.5 252,-498.5 35,-498.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_69</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"35,-521.5 114,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-498.5 114,-544.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-521.5 169,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-498.5 169,-544.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-529.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-521.5 252,-521.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-506.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "</g>\n",
              "<!-- 139796616604944&#45;&gt;139796616605184 -->\n",
              "<g id=\"edge3\" class=\"edge\">\n",
              "<title>139796616604944&#45;&gt;139796616605184</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-581.37C143.5,-573.15 143.5,-563.66 143.5,-554.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-554.61 143.5,-544.61 140,-554.61 147,-554.61\"/>\n",
              "</g>\n",
              "<!-- 139796614480176 -->\n",
              "<g id=\"node5\" class=\"node\">\n",
              "<title>139796614480176</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"40.5,-415.5 40.5,-461.5 246.5,-461.5 246.5,-415.5 40.5,-415.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_80</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"40.5,-438.5 108.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-415.5 108.5,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-438.5 163.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-415.5 163.5,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 200)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-438.5 246.5,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796616605184&#45;&gt;139796614480176 -->\n",
              "<g id=\"edge4\" class=\"edge\">\n",
              "<title>139796616605184&#45;&gt;139796614480176</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-498.37C143.5,-490.15 143.5,-480.66 143.5,-471.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-471.61 143.5,-461.61 140,-471.61 147,-471.61\"/>\n",
              "</g>\n",
              "<!-- 139796614486560 -->\n",
              "<g id=\"node6\" class=\"node\">\n",
              "<title>139796614486560</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"0,-332.5 0,-378.5 287,-378.5 287,-332.5 0,-332.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_70</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"0,-355.5 149,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-332.5 149,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"149,-355.5 204,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"176.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-332.5 204,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"204,-355.5 287,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796614480176&#45;&gt;139796614486560 -->\n",
              "<g id=\"edge5\" class=\"edge\">\n",
              "<title>139796614480176&#45;&gt;139796614486560</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-415.37C143.5,-407.15 143.5,-397.66 143.5,-388.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-388.61 143.5,-378.61 140,-388.61 147,-388.61\"/>\n",
              "</g>\n",
              "<!-- 139796614486176 -->\n",
              "<g id=\"node7\" class=\"node\">\n",
              "<title>139796614486176</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"35,-249.5 35,-295.5 252,-295.5 252,-249.5 35,-249.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_70</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"35,-272.5 114,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-249.5 114,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"114,-272.5 169,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"141.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-249.5 169,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"169,-272.5 252,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "</g>\n",
              "<!-- 139796614486560&#45;&gt;139796614486176 -->\n",
              "<g id=\"edge6\" class=\"edge\">\n",
              "<title>139796614486560&#45;&gt;139796614486176</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-332.37C143.5,-324.15 143.5,-314.66 143.5,-305.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-305.61 143.5,-295.61 140,-305.61 147,-305.61\"/>\n",
              "</g>\n",
              "<!-- 139796614487952 -->\n",
              "<g id=\"node8\" class=\"node\">\n",
              "<title>139796614487952</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"40.5,-166.5 40.5,-212.5 246.5,-212.5 246.5,-166.5 40.5,-166.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_81</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"40.5,-189.5 108.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"74.5\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-166.5 108.5,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"108.5,-189.5 163.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"136\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-166.5 163.5,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 100)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"163.5,-189.5 246.5,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"205\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614486176&#45;&gt;139796614487952 -->\n",
              "<g id=\"edge7\" class=\"edge\">\n",
              "<title>139796614486176&#45;&gt;139796614487952</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-249.37C143.5,-241.15 143.5,-231.66 143.5,-222.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-222.61 143.5,-212.61 140,-222.61 147,-222.61\"/>\n",
              "</g>\n",
              "<!-- 139796614495392 -->\n",
              "<g id=\"node9\" class=\"node\">\n",
              "<title>139796614495392</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"3.5,-83.5 3.5,-129.5 283.5,-129.5 283.5,-83.5 3.5,-83.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">batch_normalization_71</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"3.5,-106.5 152.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">BatchNormalization</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"152.5,-83.5 152.5,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"180\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"152.5,-106.5 207.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"180\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"207.5,-83.5 207.5,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"207.5,-106.5 283.5,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"245.5\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614487952&#45;&gt;139796614495392 -->\n",
              "<g id=\"edge8\" class=\"edge\">\n",
              "<title>139796614487952&#45;&gt;139796614495392</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-166.37C143.5,-158.15 143.5,-148.66 143.5,-139.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-139.61 143.5,-129.61 140,-139.61 147,-139.61\"/>\n",
              "</g>\n",
              "<!-- 139796614481664 -->\n",
              "<g id=\"node10\" class=\"node\">\n",
              "<title>139796614481664</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"38.5,-0.5 38.5,-46.5 248.5,-46.5 248.5,-0.5 38.5,-0.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">dropout_71</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"38.5,-23.5 117.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"78\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dropout</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"117.5,-0.5 117.5,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"145\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"117.5,-23.5 172.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"145\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"172.5,-0.5 172.5,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"172.5,-23.5 248.5,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"210.5\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796614495392&#45;&gt;139796614481664 -->\n",
              "<g id=\"edge9\" class=\"edge\">\n",
              "<title>139796614495392&#45;&gt;139796614481664</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M143.5,-83.37C143.5,-75.15 143.5,-65.66 143.5,-56.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"147,-56.61 143.5,-46.61 140,-56.61 147,-56.61\"/>\n",
              "</g>\n",
              "</g>\n",
              "</svg>\n",
              "</div>\n",
              "    </div>\n",
              "    <div style='width: 33%; text-align: center;'>\n",
              "        <h2>Imaging Model</h2>\n",
              "        <div><?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
              "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
              " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
              "<!-- Generated by graphviz version 2.43.0 (0)\n",
              " -->\n",
              "<!-- Title: G Pages: 1 -->\n",
              "<svg width=\"325pt\" height=\"588pt\"\n",
              " viewBox=\"0.00 0.00 260.00 470.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
              "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.8 0.8) rotate(0) translate(4 466)\">\n",
              "<title>G</title>\n",
              "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-466 256,-466 256,4 -4,4\"/>\n",
              "<!-- 139796614535424 -->\n",
              "<g id=\"node1\" class=\"node\">\n",
              "<title>139796614535424</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"0,-415.5 0,-461.5 252,-461.5 252,-415.5 0,-415.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"38.5\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">input_30</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"0,-438.5 77,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"38.5\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">InputLayer</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"77,-415.5 77,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"104.5\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"77,-438.5 132,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"104.5\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"132,-415.5 132,-461.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 72, 72, 3)]</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"132,-438.5 252,-438.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-423.3\" font-family=\"Times,serif\" font-size=\"14.00\">[(None, 72, 72, 3)]</text>\n",
              "</g>\n",
              "<!-- 139796614480320 -->\n",
              "<g id=\"node2\" class=\"node\">\n",
              "<title>139796614480320</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"1,-332.5 1,-378.5 251,-378.5 251,-332.5 1,-332.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">conv2d_30</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"1,-355.5 78,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">Conv2D</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-332.5 78,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-355.5 133,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-332.5 133,-378.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-363.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 72, 72, 3)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-355.5 251,-355.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-340.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 70, 70, 72)</text>\n",
              "</g>\n",
              "<!-- 139796614535424&#45;&gt;139796614480320 -->\n",
              "<g id=\"edge1\" class=\"edge\">\n",
              "<title>139796614535424&#45;&gt;139796614480320</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M126,-415.37C126,-407.15 126,-397.66 126,-388.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"129.5,-388.61 126,-378.61 122.5,-388.61 129.5,-388.61\"/>\n",
              "</g>\n",
              "<!-- 139796614746160 -->\n",
              "<g id=\"node3\" class=\"node\">\n",
              "<title>139796614746160</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"1,-249.5 1,-295.5 251,-295.5 251,-249.5 1,-249.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">conv2d_31</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"1,-272.5 78,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">Conv2D</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-249.5 78,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-272.5 133,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-249.5 133,-295.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-280.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 70, 70, 72)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-272.5 251,-272.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-257.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 68, 68, 64)</text>\n",
              "</g>\n",
              "<!-- 139796614480320&#45;&gt;139796614746160 -->\n",
              "<g id=\"edge2\" class=\"edge\">\n",
              "<title>139796614480320&#45;&gt;139796614746160</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M126,-332.37C126,-324.15 126,-314.66 126,-305.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"129.5,-305.61 126,-295.61 122.5,-305.61 129.5,-305.61\"/>\n",
              "</g>\n",
              "<!-- 139796614748224 -->\n",
              "<g id=\"node4\" class=\"node\">\n",
              "<title>139796614748224</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"1,-166.5 1,-212.5 251,-212.5 251,-166.5 1,-166.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">conv2d_32</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"1,-189.5 78,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">Conv2D</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-166.5 78,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"78,-189.5 133,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"105.5\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-166.5 133,-212.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-197.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 68, 68, 64)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"133,-189.5 251,-189.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"192\" y=\"-174.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 66, 66, 32)</text>\n",
              "</g>\n",
              "<!-- 139796614746160&#45;&gt;139796614748224 -->\n",
              "<g id=\"edge3\" class=\"edge\">\n",
              "<title>139796614746160&#45;&gt;139796614748224</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M126,-249.37C126,-241.15 126,-231.66 126,-222.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"129.5,-222.61 126,-212.61 122.5,-222.61 129.5,-222.61\"/>\n",
              "</g>\n",
              "<!-- 139796621685616 -->\n",
              "<g id=\"node5\" class=\"node\">\n",
              "<title>139796621685616</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"4,-83.5 4,-129.5 248,-129.5 248,-83.5 4,-83.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">flatten_10</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"4,-106.5 75,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"39.5\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">Flatten</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"75,-83.5 75,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"102.5\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"75,-106.5 130,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"102.5\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"130,-83.5 130,-129.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"189\" y=\"-114.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 66, 66, 32)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"130,-106.5 248,-106.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"189\" y=\"-91.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 139392)</text>\n",
              "</g>\n",
              "<!-- 139796614748224&#45;&gt;139796621685616 -->\n",
              "<g id=\"edge4\" class=\"edge\">\n",
              "<title>139796614748224&#45;&gt;139796621685616</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M126,-166.37C126,-158.15 126,-148.66 126,-139.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"129.5,-139.61 126,-129.61 122.5,-139.61 129.5,-139.61\"/>\n",
              "</g>\n",
              "<!-- 139796617814144 -->\n",
              "<g id=\"node6\" class=\"node\">\n",
              "<title>139796617814144</title>\n",
              "<polygon fill=\"none\" stroke=\"black\" points=\"13,-0.5 13,-46.5 239,-46.5 239,-0.5 13,-0.5\"/>\n",
              "<text text-anchor=\"middle\" x=\"47\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">dense_82</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"13,-23.5 81,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"47\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">Dense</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"81,-0.5 81,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"108.5\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">input:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"81,-23.5 136,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"108.5\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">output:</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"136,-0.5 136,-46.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"187.5\" y=\"-31.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 139392)</text>\n",
              "<polyline fill=\"none\" stroke=\"black\" points=\"136,-23.5 239,-23.5 \"/>\n",
              "<text text-anchor=\"middle\" x=\"187.5\" y=\"-8.3\" font-family=\"Times,serif\" font-size=\"14.00\">(None, 50)</text>\n",
              "</g>\n",
              "<!-- 139796621685616&#45;&gt;139796617814144 -->\n",
              "<g id=\"edge5\" class=\"edge\">\n",
              "<title>139796621685616&#45;&gt;139796617814144</title>\n",
              "<path fill=\"none\" stroke=\"black\" d=\"M126,-83.37C126,-75.15 126,-65.66 126,-56.73\"/>\n",
              "<polygon fill=\"black\" stroke=\"black\" points=\"129.5,-56.61 126,-46.61 122.5,-56.61 129.5,-56.61\"/>\n",
              "</g>\n",
              "</g>\n",
              "</svg>\n",
              "</div>\n",
              "    </div>\n",
              "</div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 56
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_classification_report(y_tru, y_prd, mode, learning_rate, batch_size, epochs, figsize=(7, 7), ax=None):\n",
        "    \"\"\" Generate and save a classification report as a heatmap. \"\"\"\n",
        "    xticks = ['precision', 'recall', 'f1-score', 'support']\n",
        "    yticks = [\"Control\", \"Moderate\", \"Alzheimer's\", 'avg']\n",
        "    rep = np.array(precision_recall_fscore_support(y_tru, y_prd)).T\n",
        "    avg = np.mean(rep, axis=0)\n",
        "    avg[-1] = np.sum(rep[:, -1])\n",
        "    rep = np.insert(rep, rep.shape[0], avg, axis=0)\n",
        "    sns.heatmap(rep, annot=True, cbar=False, xticklabels=xticks, yticklabels=yticks, ax=ax, cmap=\"Blues\")\n",
        "    plt.savefig(f'report_{mode}_{learning_rate}_{batch_size}_{epochs}.png')"
      ],
      "metadata": {
        "id": "cLn0qLyiI7Zl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_classification_report(y_tru, y_prd, mode, learning_rate, batch_size,epochs, figsize=(7, 7), ax=None):\n",
        "\n",
        "    plt.figure(figsize=figsize)\n",
        "\n",
        "    xticks = ['precision', 'recall', 'f1-score', 'support']\n",
        "    yticks = [\"Control\", \"Moderate\", \"Alzheimer's\" ]\n",
        "    yticks += ['avg']\n",
        "\n",
        "    rep = np.array(precision_recall_fscore_support(y_tru, y_prd)).T\n",
        "    avg = np.mean(rep, axis=0)\n",
        "    avg[-1] = np.sum(rep[:, -1])\n",
        "    rep = np.insert(rep, rep.shape[0], avg, axis=0)\n",
        "\n",
        "    sns.heatmap(rep,\n",
        "                annot=True,\n",
        "                cbar=False,\n",
        "                xticklabels=xticks,\n",
        "                yticklabels=yticks,\n",
        "                ax=ax, cmap = \"Blues\")\n",
        "\n",
        "    plt.savefig('report_' + str(mode) + '_' + str(learning_rate) +'_' + str(batch_size)+'_' + str(epochs)+'.png')"
      ],
      "metadata": {
        "id": "Qw1sNcvGI91j"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def calc_confusion_matrix(result, test_label,mode, learning_rate, batch_size, epochs):\n",
        "    test_label = to_categorical(test_label,3)\n",
        "\n",
        "    true_label= np.argmax(test_label, axis =1)\n",
        "\n",
        "    predicted_label= np.argmax(result, axis =1)\n",
        "\n",
        "    n_classes = 3\n",
        "    precision = dict()\n",
        "    recall = dict()\n",
        "    thres = dict()\n",
        "    for i in range(n_classes):\n",
        "        precision[i], recall[i], thres[i] = precision_recall_curve(test_label[:, i],\n",
        "                                                            result[:, i])\n",
        "\n",
        "\n",
        "    print (\"Classification Report :\")\n",
        "    print (classification_report(true_label, predicted_label))\n",
        "    cr = classification_report(true_label, predicted_label, output_dict=True)\n",
        "    return cr, precision, recall, thres"
      ],
      "metadata": {
        "id": "bcJ1iiqyQRoX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Attention/Multi-modal Model\n",
        "\n",
        "def cross_modal_attention(x, y):\n",
        "    x = tf.expand_dims(x, axis=1)\n",
        "    y = tf.expand_dims(y, axis=1)\n",
        "    a1 = MultiHeadAttention(num_heads = 4,key_dim=50)(x, y)\n",
        "    a2 = MultiHeadAttention(num_heads = 4,key_dim=50)(y, x)\n",
        "    a1 = a1[:,0,:]\n",
        "    a2 = a2[:,0,:]\n",
        "    return concatenate([a1, a2])\n",
        "\n",
        "\n",
        "def self_attention(x):\n",
        "    x = tf.expand_dims(x, axis=1)\n",
        "    attention = MultiHeadAttention(num_heads = 4, key_dim=50)(x, x)\n",
        "    attention = attention[:,0,:]\n",
        "    return attention\n",
        "\n",
        "\n",
        "def multi_modal_model(mode, train_clinical, train_snp, train_img):\n",
        "\n",
        "    in_clinical = Input(shape=(train_clinical.shape[1]))\n",
        "\n",
        "    in_snp = Input(shape=(train_snp.shape[1]))\n",
        "\n",
        "    in_img = Input(shape=(train_img.shape[1], train_img.shape[2], train_img.shape[3]))\n",
        "\n",
        "    dense_clinical = create_model_clinical()(in_clinical)\n",
        "    dense_snp = create_model_snp()(in_snp)\n",
        "    dense_img = create_model_img()(in_img)\n",
        "\n",
        "    ########### Attention Layer ############\n",
        "\n",
        "    ## Cross Modal Bi-directional Attention ##\n",
        "\n",
        "    if mode == 'MM_BA':\n",
        "\n",
        "        vt_att = cross_modal_attention(dense_img, dense_clinical)\n",
        "        av_att = cross_modal_attention(dense_snp, dense_img)\n",
        "        ta_att = cross_modal_attention(dense_clinical, dense_snp)\n",
        "\n",
        "        merged = concatenate([vt_att, av_att, ta_att, dense_img, dense_snp, dense_clinical])\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "    ## Self Attention ##\n",
        "    elif mode == 'MM_SA':\n",
        "\n",
        "        vv_att = self_attention(dense_img)\n",
        "        tt_att = self_attention(dense_clinical)\n",
        "        aa_att = self_attention(dense_snp)\n",
        "\n",
        "        merged = concatenate([aa_att, vv_att, tt_att, dense_img, dense_snp, dense_clinical])\n",
        "\n",
        "    ## Self Attention and Cross Modal Bi-directional Attention##\n",
        "    elif mode == 'MM_SA_BA':\n",
        "\n",
        "        vv_att = self_attention(dense_img)\n",
        "        tt_att = self_attention(dense_clinical)\n",
        "        aa_att = self_attention(dense_snp)\n",
        "\n",
        "        vt_att = cross_modal_attention(vv_att, tt_att)\n",
        "        av_att = cross_modal_attention(aa_att, vv_att)\n",
        "        ta_att = cross_modal_attention(tt_att, aa_att)\n",
        "\n",
        "        merged = concatenate([vt_att, av_att, ta_att, dense_img, dense_snp, dense_clinical])\n",
        "\n",
        "\n",
        "    ## No Attention ##\n",
        "    elif mode == 'None':\n",
        "\n",
        "        merged = concatenate([dense_img, dense_snp, dense_clinical])\n",
        "\n",
        "    else:\n",
        "        print (\"Mode must be one of 'MM_SA', 'MM_BA', 'MU_SA_BA' or 'None'.\")\n",
        "        return\n",
        "\n",
        "\n",
        "    ########### Output Layer ############\n",
        "\n",
        "    output = Dense(3, activation='softmax')(merged)\n",
        "    model = Model([in_clinical, in_snp, in_img], output)\n",
        "\n",
        "    return model"
      ],
      "metadata": {
        "id": "1X3dVQNuQZb8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Training\n",
        "\n",
        "### Hyperparameter Tuning\n",
        "\n",
        "#### Batch Size\n",
        "The batch size was tuned. The best batch size in the original paper was 32.\n",
        "\n",
        "#### Epochs\n",
        "The number of training epochs was tuned. The best performing number of epochs in the original paper was 50\n",
        "\n",
        "#### Learning Rate\n",
        "The learning rate was tuned. The best performing learning rate in the original paper was .001\n",
        "\n",
        "#### Seed\n",
        "The random seed was was tuned in a range of 0-200 in steps of 5.\n",
        "\n",
        "\n",
        "### Computational Requirements and Considerations:\n",
        "#### Hardware\n",
        "Using a T4 GPU hardware accelerator or another GPU type accellerator is necessary to process the large amount of imaging data as well as the extremely large VCF files for the genetic data.\n",
        "\n",
        "#### Runtime\n",
        "The runtime would vary quite a bit per epoch, but the smaller batch size and relatively smaller learning rate would reduce the training time per epoch by a significant amount."
      ],
      "metadata": {
        "id": "PBI-wBgM7xv5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def train(mode, batch_size, epochs, learning_rate, seed):\n",
        "\n",
        "\n",
        "    train_clinical = pd.read_csv(\"X_train_clinical.csv\").drop(\"Unnamed: 0\", axis=1).values\n",
        "    test_clinical= pd.read_csv(\"X_test_clinical.csv\").drop(\"Unnamed: 0\", axis=1).values\n",
        "\n",
        "\n",
        "    train_snp = pd.read_csv(\"X_train_snp.csv\").drop(\"Unnamed: 0\", axis=1).values\n",
        "    test_snp = pd.read_csv(\"X_test_snp.csv\").drop(\"Unnamed: 0\", axis=1).values\n",
        "\n",
        "\n",
        "    train_img= make_img(\"X_train_img.pkl\")\n",
        "    test_img= make_img(\"X_test_img.pkl\")\n",
        "\n",
        "\n",
        "    train_label= pd.read_csv(\"y_train.csv\").drop(\"Unnamed: 0\", axis=1).values.astype(\"int\").flatten()\n",
        "    test_label= pd.read_csv(\"y_test.csv\").drop(\"Unnamed: 0\", axis=1).values.astype(\"int\").flatten()\n",
        "\n",
        "    reset_random_seeds(seed)\n",
        "    class_weights = compute_class_weight(class_weight = 'balanced',classes = np.unique(train_label),y = train_label)\n",
        "    d_class_weights = dict(enumerate(class_weights))\n",
        "\n",
        "    # compile model #\n",
        "    model = multi_modal_model(mode, train_clinical, train_snp, train_img)\n",
        "    model.compile(optimizer=Adam(learning_rate = learning_rate), loss='sparse_categorical_crossentropy', metrics=['sparse_categorical_accuracy'])\n",
        "\n",
        "\n",
        "    # summarize results\n",
        "    history = model.fit([train_clinical,\n",
        "                         train_snp,\n",
        "                         train_img],\n",
        "                        train_label,\n",
        "                        epochs=epochs,\n",
        "                        batch_size=batch_size,\n",
        "                        class_weight=d_class_weights,\n",
        "                        validation_split=0.1,\n",
        "                        verbose=1)\n",
        "\n",
        "\n",
        "\n",
        "    score = model.evaluate([test_clinical, test_snp, test_img], test_label)\n",
        "\n",
        "    acc = score[1]\n",
        "    test_predictions = model.predict([test_clinical, test_snp, test_img])\n",
        "    cr, precision_d, recall_d, thres = calc_confusion_matrix(test_predictions, test_label, mode, learning_rate, batch_size, epochs)\n",
        "\n",
        "\n",
        "    \"\"\"\n",
        "    plt.clf()\n",
        "    plt.plot(history.history['sparse_categorical_accuracy'])\n",
        "    plt.plot(history.history['val_sparse_categorical_accuracy'])\n",
        "    plt.title('model accuracy')\n",
        "    plt.ylabel('accuracy')\n",
        "    plt.xlabel('epoch')\n",
        "    plt.legend(['train', 'validation'], loc='upper left')\n",
        "    plt.show()\n",
        "    plt.savefig('accuracy_' + str(mode) + '_' + str(learning_rate) +'_' + str(batch_size)+'.png')\n",
        "    plt.clf()\n",
        "    # summarize history for loss\n",
        "    plt.plot(history.history['loss'])\n",
        "    plt.plot(history.history['val_loss'])\n",
        "    plt.title('model loss')\n",
        "    plt.ylabel('loss')\n",
        "    plt.xlabel('epoch')\n",
        "    plt.legend(['train', 'validation'], loc='upper left')\n",
        "    plt.show()\n",
        "    plt.savefig('loss_' + str(mode) + '_' + str(learning_rate) +'_' + str(batch_size)+'.png')\n",
        "    plt.clf()\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "\n",
        "    # release gpu memory #\n",
        "    K.clear_session()\n",
        "    del model, history\n",
        "    gc.collect()\n",
        "\n",
        "\n",
        "    print ('Mode: ', mode)\n",
        "    print ('Batch size:  ', batch_size)\n",
        "    print ('Learning rate: ', learning_rate)\n",
        "    print ('Epochs:  ', epochs)\n",
        "    print ('Test Accuracy:', '{0:.4f}'.format(acc))\n",
        "    print ('-'*55)\n",
        "\n",
        "    return acc, batch_size, learning_rate, epochs, seed"
      ],
      "metadata": {
        "id": "mwZWfDBKQrDu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "    m_a = {}\n",
        "    seeds = random.sample(range(1, 200), 5)\n",
        "    for s in seeds:\n",
        "        acc, bs_, lr_, e_ , seed= train('MM_SA_BA', 32, 50, 0.001, s)\n",
        "        m_a[acc] = ('MM_SA_BA', acc, bs_, lr_, e_, seed)\n",
        "    print(m_a)\n",
        "    print ('-'*55)\n",
        "    max_acc = max(m_a, key=float)\n",
        "    print(\"Highest accuracy of: \" + str(max_acc) + \" with parameters: \" + str(m_a[max_acc]))"
      ],
      "metadata": {
        "id": "hDo-z8hIJC1r"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Results\n",
        "\n",
        "The training of the MADDi model experienced several challenges that proved fatal to the reproducibility of the original paper.\n",
        "\n",
        "Training was not able to be completed due to these challenges.\n",
        "\n",
        "Expected outcomes are included below."
      ],
      "metadata": {
        "id": "gX6bCcZNuxmz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "## Expected Outcomes\n",
        "\n",
        "#Upon completion model training, we would expect the to match or improve upon the following classification results.\n",
        "\n",
        "import pandas as pd\n",
        "from IPython.display import display\n",
        "\n",
        "# Create a DataFrame with the performance metrics\n",
        "data = {\n",
        "    'Category': ['Control', 'Moderate Cognitive Impairment', 'Alzheimer’s Disease'],\n",
        "    'Accuracy (%)': [96.66, 96.66, 100],\n",
        "    'Precision (%)': [96.78, 90.00, 100],\n",
        "    'Recall (%)': [98.88, 70.00, 100],\n",
        "    'F1-Score (%)': [97.81, 76.66, 100]\n",
        "}\n",
        "\n",
        "df = pd.DataFrame(data)\n",
        "\n",
        "display(df)"
      ],
      "metadata": {
        "id": "LjW9bCkouv8O",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 143
        },
        "outputId": "7b170c0a-84f5-4d7d-ca7f-e05708eb6db4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                        Category  Accuracy (%)  Precision (%)  Recall (%)  \\\n",
              "0                        Control         96.66          96.78       98.88   \n",
              "1  Moderate Cognitive Impairment         96.66          90.00       70.00   \n",
              "2            Alzheimer’s Disease        100.00         100.00      100.00   \n",
              "\n",
              "   F1-Score (%)  \n",
              "0         97.81  \n",
              "1         76.66  \n",
              "2        100.00  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b838c50b-8c8b-4748-b77a-ef25845bb8d2\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Category</th>\n",
              "      <th>Accuracy (%)</th>\n",
              "      <th>Precision (%)</th>\n",
              "      <th>Recall (%)</th>\n",
              "      <th>F1-Score (%)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Control</td>\n",
              "      <td>96.66</td>\n",
              "      <td>96.78</td>\n",
              "      <td>98.88</td>\n",
              "      <td>97.81</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Moderate Cognitive Impairment</td>\n",
              "      <td>96.66</td>\n",
              "      <td>90.00</td>\n",
              "      <td>70.00</td>\n",
              "      <td>76.66</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Alzheimer’s Disease</td>\n",
              "      <td>100.00</td>\n",
              "      <td>100.00</td>\n",
              "      <td>100.00</td>\n",
              "      <td>100.00</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b838c50b-8c8b-4748-b77a-ef25845bb8d2')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b838c50b-8c8b-4748-b77a-ef25845bb8d2 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b838c50b-8c8b-4748-b77a-ef25845bb8d2');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-2b4ca3ba-691d-4281-bd9d-d4af77455884\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-2b4ca3ba-691d-4281-bd9d-d4af77455884')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-2b4ca3ba-691d-4281-bd9d-d4af77455884 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 3,\n  \"fields\": [\n    {\n      \"column\": \"Category\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"Control\",\n          \"Moderate Cognitive Impairment\",\n          \"Alzheimer\\u2019s Disease\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Accuracy (%)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1.928349899093352,\n        \"min\": 96.66,\n        \"max\": 100.0,\n        \"num_unique_values\": 2,\n        \"samples\": [\n          100.0,\n          96.66\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Precision (%)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5.104520872063639,\n        \"min\": 90.0,\n        \"max\": 100.0,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          96.78,\n          90.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Recall (%)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 17.00641447611263,\n        \"min\": 70.0,\n        \"max\": 100.0,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          98.88,\n          70.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"F1-Score (%)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 12.889751743148512,\n        \"min\": 76.66,\n        \"max\": 100.0,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          97.81,\n          76.66\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Ablations\n",
        "\n",
        "#### Model Composition\n",
        "Ablations were performed in the original paper to find the ideal model composition of MADDi. Models were trained with different \"modes\".\n",
        "Different architecture features were ablated to determine the importance of the different data modalities.\n",
        "\n",
        "The best performing model was **Self Attention and Cross Modal Bi-directional Attention** in the original paper.\n",
        "#### Mode Types\n",
        "\n",
        "*  **Cross Modal Bi-directional Attention**\n",
        "*  **Self Attention**\n",
        "*  **Self Attention and Cross Modal Bi-directional Attention**\n",
        "*  **General concatenated model, no attention**"
      ],
      "metadata": {
        "id": "L9tHK890HnHJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Model comparison"
      ],
      "metadata": {
        "id": "8EAWAy_LwHlV"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "A comparison is given using the original paper's results in the absence of reproduced results in this project."
      ],
      "metadata": {
        "id": "Taj2FdDuMWhH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from IPython.display import display\n",
        "\n",
        "comparison_data = {\n",
        "    'Study': [\n",
        "        'Bucholc et al, 2019',\n",
        "        'Fang et al, 2020',\n",
        "        'Abuhmed et al, 2021',\n",
        "        'Venugopalan et al, 2021',\n",
        "        'MADDi, Original Paper'\n",
        "    ],\n",
        "    'Modality': [\n",
        "        'MRI, PET, Clinical',\n",
        "        'MRI, PET',\n",
        "        'MRI, PET, Clinical',\n",
        "        'MRI, SNP, Clinical',\n",
        "        'MRI, SNP, Clinical'\n",
        "    ],\n",
        "    'Accuracy (%)': [\n",
        "        82.90,\n",
        "        66.29,\n",
        "        86.08,\n",
        "        78.00,\n",
        "        96.88\n",
        "    ],\n",
        "    'F1-Score (%)': [\n",
        "        'Not reported',\n",
        "        'Not reported',\n",
        "        87.67,\n",
        "        78.00,\n",
        "        91.41\n",
        "    ],\n",
        "    'Method': [\n",
        "        'SVM',\n",
        "        'GDCA',\n",
        "        'Multivariate BiLSTM',\n",
        "        'DL + RF',\n",
        "        'DL + Attention'\n",
        "    ]\n",
        "}\n",
        "\n",
        "comparison_df = pd.DataFrame(comparison_data)\n",
        "\n",
        "display(comparison_df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "WtrFOFHOMUoX",
        "outputId": "46a44b97-709b-4f9f-889a-6147cbf5a674"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                     Study            Modality  Accuracy (%)  F1-Score (%)  \\\n",
              "0      Bucholc et al, 2019  MRI, PET, Clinical         82.90  Not reported   \n",
              "1         Fang et al, 2020            MRI, PET         66.29  Not reported   \n",
              "2      Abuhmed et al, 2021  MRI, PET, Clinical         86.08         87.67   \n",
              "3  Venugopalan et al, 2021  MRI, SNP, Clinical         78.00          78.0   \n",
              "4    MADDi, Original Paper  MRI, SNP, Clinical         96.88         91.41   \n",
              "\n",
              "                Method  \n",
              "0                  SVM  \n",
              "1                 GDCA  \n",
              "2  Multivariate BiLSTM  \n",
              "3              DL + RF  \n",
              "4       DL + Attention  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-86fad482-8fb9-4ce8-b706-61684c7bec48\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Study</th>\n",
              "      <th>Modality</th>\n",
              "      <th>Accuracy (%)</th>\n",
              "      <th>F1-Score (%)</th>\n",
              "      <th>Method</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Bucholc et al, 2019</td>\n",
              "      <td>MRI, PET, Clinical</td>\n",
              "      <td>82.90</td>\n",
              "      <td>Not reported</td>\n",
              "      <td>SVM</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Fang et al, 2020</td>\n",
              "      <td>MRI, PET</td>\n",
              "      <td>66.29</td>\n",
              "      <td>Not reported</td>\n",
              "      <td>GDCA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Abuhmed et al, 2021</td>\n",
              "      <td>MRI, PET, Clinical</td>\n",
              "      <td>86.08</td>\n",
              "      <td>87.67</td>\n",
              "      <td>Multivariate BiLSTM</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Venugopalan et al, 2021</td>\n",
              "      <td>MRI, SNP, Clinical</td>\n",
              "      <td>78.00</td>\n",
              "      <td>78.0</td>\n",
              "      <td>DL + RF</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>MADDi, Original Paper</td>\n",
              "      <td>MRI, SNP, Clinical</td>\n",
              "      <td>96.88</td>\n",
              "      <td>91.41</td>\n",
              "      <td>DL + Attention</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-86fad482-8fb9-4ce8-b706-61684c7bec48')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-86fad482-8fb9-4ce8-b706-61684c7bec48 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-86fad482-8fb9-4ce8-b706-61684c7bec48');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-65b21423-9ee1-457d-b259-39abb684626f\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-65b21423-9ee1-457d-b259-39abb684626f')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-65b21423-9ee1-457d-b259-39abb684626f button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "comparison_df",
              "summary": "{\n  \"name\": \"comparison_df\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"Study\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"Fang et al, 2020\",\n          \"MADDi, Original Paper\",\n          \"Abuhmed et al, 2021\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Modality\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"MRI, PET, Clinical\",\n          \"MRI, PET\",\n          \"MRI, SNP, Clinical\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Accuracy (%)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 11.198999955353152,\n        \"min\": 66.29,\n        \"max\": 96.88,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          66.29,\n          96.88,\n          86.08\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"F1-Score (%)\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          87.67,\n          91.41,\n          \"Not reported\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Method\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"GDCA\",\n          \"DL + Attention\",\n          \"Multivariate BiLSTM\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Discussion\n",
        "\n",
        "### Reproducibility\n",
        "The paper was not found to be easily reproducible.\n",
        "\n",
        "#### Barriers to reproducibility\n",
        "- **Data Access:** The ADNI data is not publicly accessible and the process of downloading and working with the datasets involved in the paper are cumbersome. The genetic data is extremely large when downloaded, but is then processed down to a more manageable size.\n",
        "\n",
        "- **Imprecise, Outdated Data Descriptions**: The original paper's description of the datasets used was extremely imprecise and out of date. There is a lot of guess work involved in finding the proper datasets to train MADDi with. Since the training of the original paper's model, ADNI3 was completed and ADNI4 data is also available, crowding the data archives of LONI. Several Github issues with questions about the data used were opened in the original repo, with vague responses by the original researcher. The non-public nature of the dataset may have led the researcher to be overly cautious when sharing descriptions of the data, as well as the changing nature of the LONI website itself.\n",
        "\n",
        "- **Data Size**: Without precise data instructions from the original paper, it is necessary to download large swathes of data which can then be preprocessed and pared down to the size of the original dataset. VCF files are extremely large, and so are the images in the MRI datasets. This introduces a complexity when developing in a Google Colab environment, where memory is limited and the transfer of data between Google Drives is slow.\n",
        "\n",
        "#### Conducive to reproducibility\n",
        "- **Model Architecture:** The model architecture is now extremely popular with the rise of transformers and attention, making the structure of the model involved easy to copy and improve upon with the range of research available to draw on.\n",
        "\n",
        "### Next Phase\n",
        "The next phase is to complete the training and perform additional ablations using the now complete ADNI3 dataset.\n",
        "\n",
        "### Recommendations for Increasing Reproducibility\n",
        "- **Detailed Descriptions of Data Acquisition** - adding more detailed descriptions of how the particular datasets were acquired within LONI would make reproducing MADDi extremely simple. The ADNI datasets within LONI are all named, or have defining features, and there are named filters as well. Instead of providing vague directions on which filters should be used, the explicit names of the filters can be provided to simplify the process.\n"
      ],
      "metadata": {
        "id": "qH75TNU71eRH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# References\n",
        "\n",
        "Alzheimer’s Association. 2023. 2024 alzheimer’s disease facts and figures. https://www.alz.\n",
        "org/media/Documents/alzheimers-facts-and-figures.pdf. Accessed on 2024-03-24.\n",
        "https://www.alz.org/media/Documents/alzheimers-facts-and-figures.pdf.\n",
        "\n",
        "Michal Golovanevsky, Carsten Eickhoff, and Ritambhara Singh. 2022. Multimodal attention-based\n",
        "deep learning for Alzheimer’s disease diagnosis. Journal of the American Medical Informatics\n",
        "Association 29(12):2014–2022. https://doi.org/10.1093/jamia/ocac168.\n",
        "\n"
      ],
      "metadata": {
        "id": "SHMI2chl9omn"
      }
    }
  ]
}